### ISSUE TITLE
Allow for complete override of GUI-set QEMU arguments

### ISSUE THREAD

**Issue by CONIGUERO** [utmapp/UTM#4102]
There are times in which it's necessary to add parameters to the "core" qemu arguments apart from the "machine" one, for example with `-device vga,edid=on` and many others.

Many of these optional params are not available in the settings GUI, and since the core arguments are locked down to the options specified in the VM settings it's impossible to ever run a vm with them.

I propose the following solution: 
* Let the user customize **all** the QEMU arguments through the QEMU tab
* Add up/down arrows on all the arguments so that the user can easily customize the order as if they were writing it on the command line
* When an argument that doesn't have a GUI equivalent is detected, add a "Custom" label to the corresponding UI dropdown menu/box in the settings panel to indicate that the argument/setting is currently mapped to a custom value that doesn't have a GUI equivalent yet. 
* Allow the user to override their custom arguments by simply selecting a known configuration parameter from the GUI again.

**Comment by osy** [comment]
So this is exactly what “Do not generate any arguments based on current configuration” does.

**Comment by osy** [comment]
Can you describe your use case for “-device vga,edid=on”? If it’s a valid use case, it should be added to the GUI.

**Comment by CONIGUERO** [comment]
> So this is exactly what “Do not generate any arguments based on current configuration” does.

AFAIK that option doesn't allow you to modify any of the core arguments, does it?

**Comment by CONIGUERO** [comment]
> Can you describe your use case for “-device vga,edid=on”? If it’s a valid use case, it should be added to the GUI.

That was just an example. One valid use case for such options is to disable auto-boot on PowerPC guests using the `-prom-env 'auto-boot?=false'` argument.
In order to successfully install Mac OS X developer preview 3, you need to select a specific partition inside the ISO to boot from, because the default one boots into an incompatible environment.

**Comment by CONIGUERO** [comment]
> For that one, I think adding a new argument is fine. The idea is anything that is "core" (like VGA settings) should be a GUI option and if there's a valid use case there, it should be added to the GUI rather than as a text argument.

I also agree that it should be made into a GUI option. My point is that for more advanced use cases that don't yet (or ever) have GUI options it should be allowed for the user to modify them as if it was a command-line invocation. And for the ones that do already have a GUI, manual modification should also still be allowed as an "advanced" mode, since the current configuration override doesn't cover for the options in the beginning of the command and as I said there are many valid use-cases for modifying them.

It could also help with development since it would be easier to test different QEMU options faster.

**Comment by osy** [comment]
So we have `-machine` property as its own field. We could add something similar to the `-device vga`. However, like I said, most of the other options you can't change.

`qemu-system-x86_64 -L /Applications/UTM.app/Contents/Resources/qemu -S -spice "unix=on,addr=....spice,disable-ticketing=on,image-compression=off,playback-compression=off,streaming-video=off,gl=off" -chardev spiceport,id=org.qemu.monitor.qmp,name=org.qemu.monitor.qmp.0 -mon chardev=org.qemu.monitor.qmp,mode=control -nodefaults -vga none -device virtio-vga`

If you look at these options, none of them except `-device virtio-vga` has anything you can add to or change.

**Comment by CONIGUERO** [comment]
I guess that's right. I was also thinking of the `-g` option which allows you to set a static resolution for the machine display. I appended it through the additional options menu but it didn't have any effect still.

**Comment by tomba4** [comment]
I have a similar issue and it's extremely annoying... E.g. QEMU emulates a network card by default for PowerPC (G4 at least) and doesn't require any parameters. When I set up UTM with no network options though, it passes "-nic none" to qemu, which is simply wrong.

Also, when Mac99 is selected, UTM adds via=pmu, which is wrong, at least for early MacOS X guests.

Etc.

The "Export QEMU command" option is interesting as it could be edited, but how to actually run it? qemu-system commands are not available from the terminal.

**Comment by eugenialoli** [comment]
I would like UTM to let us edit that qemu command line. I had multiple OSes that simply were not possible to run on UTM because it wouldn't let us change the qemu line. Some of these include the latest GNU Hurd (it's hardcoded to hd0 disks, but UTM would only produce sd0 disks no matter which type of disk chosen, and so booting would fail), and Raspbian with hypervisor (it complains about EL3 being On, and that it can't find PCI to use the network card).

**Comment by anishp55** [comment]
i really would like to be able to tell UTM what network device to use, i currently cannot figure out how to do this.  i have 3 different network connections, 1 wifi for general. 1 ethernet for my home network and device, and another that is for my lab stuff.  I would like to assign this VM to my Lab network and its currently impossible with UTM from what i can tell.

**Comment by osy** [comment]
Select bridged and select your network adapter.

**Comment by jlforester** [comment]
@osy That works if the GUI shows you the interface you want to add.  If not, you're out of luck.  In my case, when configuring bridged mode, I get two choices: `Automatic` and `en0`.  The interface I need to use doesn't show up in the list.

I don't understand the resistance to adding an Edit button and letting users edit the options to their heart's content with a stern warning that doing so is unsupported and they could render their VM unusable.  You could also provide a "Rescue" button to regenerate the command line options from the config.  Not every user needs or wants to be protected from themselves.

**Comment by sentient06** [comment]
I agree. I was trying to add `-device VGA,edid=on` to the list. I know it works and I use it outside of UTM. But I can't override the display settings on UTM. I tried to remove the display settings entirely using the GUI and re-add manually, and that crashed the application. What gives? Why is this so complicated?

**Comment by BlueFalconHD** [comment]
This would also be helpful for debugging VMs or having remote debugging servers on them
